src/vote_graph.rs

```diff
@@ -0,0 +1,515 @@
+				.expect("node either base or referenced by other in graph; qed")
+		};
+
+		let mut node_key = current_best
+			.and_then(|(hash, number)| match self.find_containing_nodes(hash.clone(), number) {
+				None => Some(hash),
+				Some(ref x) if !x.is_empty() => Some(x[0].clone()),
+				Some(_) => None,
+			})
+			.and_then(|hash| get_node(&hash).ancestor_node())
+			.unwrap_or_else(|| self.base.clone());
+
+		let mut active_node = get_node(&node_key);
+
+		if !condition(&active_node.cumulative_vote) { return None }
+
+		// breadth-first search starting from this node.
+		loop {
+			let next_descendent = active_node.descendents
+				.iter()
+				.map(|d| (d.clone(), get_node(d)))
+				.filter(|&(_, ref node)| condition(&node.cumulative_vote))
+				.next();
+
+			match next_descendent {
+				Some((key, node)) => {
+					node_key = key;
+					active_node = node;
+				}
+				None => break,
+			}
+		}
+
+		// active_node and node_key now correspond to the head node which has enough cumulative votes.
+		// we have a frontier of vote-nodes which individually don't have enough votes
+		// to pass the threshold but some subset of them join either at `active_node` or at some
+		// descendent block of it, giving that block sufficient votes.
+		Some(self.ghost_find_merge_point(node_key, active_node, condition))
+	}
+
+	// given a key, node pair (which must correspond), assuming this node fulfills the condition,
+	// this function will find the highest point at which its descendents merge, which may be the
+	// node itself.
+	fn ghost_find_merge_point<'a, F>(&'a self, mut node_key: H, mut active_node: &'a Entry<H, V>, condition: F) -> (H, usize)
+		where F: Fn(&V) -> bool
+	{
+		let mut descendent_nodes: Vec<_> = active_node.descendents.iter()
+			.map(|h| self.entries.get(h).expect("descendents always present in node storage; qed"))
+			.collect();
+
+		let base_number = active_node.number;
+		let (mut best_hash, mut best_number) = (node_key, active_node.number);
+		let mut descendent_blocks = Vec::with_capacity(descendent_nodes.len());
+
+		// TODO: for long ranges of blocks this could get inefficient
+		for offset in 1usize.. {
+			let mut new_best = None;
+			for d_node in descendent_nodes.iter() {
+				if let Some(d_block) = d_node.ancestor_block(base_number + offset) {
+					match descendent_blocks.binary_search_by_key(&d_block, |&(ref x, _)| x) {
+						Ok(idx) => {
+							descendent_blocks[idx].1 += d_node.cumulative_vote.clone();
+							if condition(&descendent_blocks[idx].1) {
+								new_best = Some(d_block.clone());
+							}
+						}
+						Err(idx) => descendent_blocks.insert(idx, (
+							d_block.clone(),
+							d_node.cumulative_vote.clone()
+						)),
+					}
+				}
+			}
+
+			match new_best {
+				Some(new_best) => {
+					best_hash = new_best;
+					best_number += 1;
+				}
+				None => break,
+			}
+
+			descendent_blocks.clear();
+			descendent_nodes.retain(
+				|n| n.in_direct_ancestry(&best_hash, best_number).unwrap_or(false)
```

Comment:
Shouldn't this be `unwrap_or(true)`? We're removing nodes that are too high compared to the current best GHOST node, i.e. there are other vote nodes in the middle. It doesn't necessarily mean that they're not part of this branch.

Reply:
If we get `None` back, then we've exhausted all the blocks in the "edge" between our GHOST vote-node and the descendent. 

Since we're looking for an ancestor-block of the descendent vote-node's block which has enough cumulative votes, once we've reached this point we know that it's safe to remove the descendent node from our frontier, as we've checked all ancestors of the descendent vote-node already.

The other theoretical case (which cannot occur in this part of the code) where `in_direct_ancestry` returns `None` is where the block is _earlier_ than the best GHOST vote-node. But we don't care about that case anyway since we're only looking for descendent blocks of that node.